File: sol/solution/0301-0400/0301/Solution.rb
def remove_invalid_parentheses(s)
  return [s] if valid?(s)
  result = []
  visited = Set.new([s])
  queue = [s]
  found = false
  until queue.empty? || found
    size = queue.length
    size.times do
      current = queue.shift
      (0...current.length).each do |i|
        next unless current[i] == '(' || current[i] == ')'
        new_str = current[0...i] + current[i+1..-1]
        next if visited.include?(new_str)
        visited.add(new_str)
        if valid?(new_str)
          result << new_str
          found = true
        else
          queue << new_str unless found
        end
      end
    end
  end
  result.empty? ? [""] : result
end
def valid?(s)
  count = 0
  s.each_char do |char|
    if char == '('
      count += 1
    elsif char == ')'
      count -= 1
      return false if count < 0
    end
  end
  count == 0
end

File: sol/solution/0301-0400/0301/Solution.ex
defmodule Solution do
  @spec remove_invalid_parentheses(s :: String.t) :: [String.t]
  def remove_invalid_parentheses(s) do
    bfs([s], MapSet.new(), [])
  end
  defp bfs([], _visited, results), do: results
  defp bfs(queue, visited, results) do
    {new_results, found, new_queue, new_visited} =
      Enum.reduce(queue, {results, false, [], visited}, fn cur, {res_acc, found_flag, nq, vis} ->
        cond do
          valid?(cur) ->
            {[cur | res_acc], true, nq, vis}
          found_flag ->
            {res_acc, true, nq, vis}
          true ->
            next_states =
              for i <- 0..(String.length(cur) - 1),
                  String.at(cur, i) in ["(", ")"] do
                String.slice(cur, 0, i) <> String.slice(cur, i + 1, String.length(cur) - i - 1)
              end
            {
              res_acc,
              found_flag,
              nq ++ Enum.filter(next_states, &(!MapSet.member?(vis, &1))),
              Enum.reduce(next_states, vis, &MapSet.put(&2, &1))
            }
        end
      end)
    if found do
      Enum.uniq(new_results)
    else
      bfs(new_queue, new_visited, new_results)
    end
  end
  defp valid?(str) do
    str
    |> String.graphemes()
    |> Enum.reduce_while(0, fn
      "(", count -> {:cont, count + 1}
      ")", count -> if count > 0, do: {:cont, count - 1}, else: {:halt, -1}
      _, count -> {:cont, count}
    end)
    |> Kernel.==(0)
  end
end
